### PR TITLE
Airlock Switch recreated.

### PR DESCRIPTION
**Pull request by Adngel**
Added the LOCK_SWITCH object in the unused AIRLOCK_SWITCH slot. Used for the valves that are designed to trigger doors.

- No ocb requried.
- No asset fixes required. (Tacking asset from TR5 submarine).

Use:
- Place the lock_switch and the door in the same sector.
- Put a Switch trigger for the lock_switch in the sector.
- Put a normal trigger for the door in the sector.

[Airlock Switch Video](https://www.youtube.com/watch?v=0i7i8ehkIlg)

### CODE REVIEW

File: TombEngine/Game/control/trigger.cpp [hunk 1]
```diff
@@ -152,6 +152,10 @@ int SwitchTrigger(short itemNumber, short timer)
 	}
 	else if (item->Status)
 	{
+		if (item->ObjectNumber == ID_AIRLOCK_SWITCH &&
+			item->Animation.AnimNumber == GetAnimNumber (*item, 2) &&
+			item->Animation.FrameNumber == GetFrameNumber (item, 0))
+			return 1;
```
Comment (Sezzary) [inline TombEngine/Game/control/trigger.cpp:158]: Good opportunity to modify `SwitchTrigger()` to return a `bool` instead of 0 or 1.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#include "framework.h"
```
Comment (Sezzary) [inline TombEngine/Objects/Generic/Switches/airlock_switch.cpp:1]: Please alphabetise includes. :)
Also, since this is a new file, it can use Pascal case: `AirlockSwitch.cpp/h`.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	};
+
+	auto AirLockSwitchPos = Vector3i::Zero;
+
+	void AirLockSwitchCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto* laraInfo = GetLaraInfo(laraItem);
+		auto* switchItem = &g_Level.Items[itemNumber];
+
+		if (TrInput & IN_ACTION &&
```
Comment (Sezzary) [inline TombEngine/Objects/Generic/Switches/airlock_switch.cpp:32]: Can use `IsHeld(In::Action)` here.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+			switchItem->Animation.ActiveState == 0 &&
+			laraItem->Animation.ActiveState == LS_IDLE &&
+			laraItem->Animation.AnimNumber == LA_STAND_IDLE &&
+			laraInfo->Control.HandStatus == HandStatus::Free &&
+			switchItem->Status == ITEM_NOT_ACTIVE ||
+			laraInfo->Control.IsMoving && laraInfo->InteractedItem == itemNumber)
+		{
+			auto bounds = GameBoundingBox(switchItem);
+
+			LockSwitchBounds.BoundingBox.X1 = bounds.X1 - 256;
```
Comment (Sezzary) [inline TombEngine/Objects/Generic/Switches/airlock_switch.cpp:42]: Offsets can be demagicked with `BLOCK(0.25f)` and `BLOCK(0.5f)`.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			EulerAngles(ANGLE(-10.0f), ANGLE(-30.0f), ANGLE(-10.0f)),
+			EulerAngles(ANGLE(10.0f), ANGLE(30.0f), ANGLE(10.0f))
+		)
+	};
+
+	auto AirLockSwitchPos = Vector3i::Zero;
+
+	void AirLockSwitchCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto* laraInfo = GetLaraInfo(laraItem);
```
Comment (Sezzary) [inline TombEngine/Objects/Generic/Switches/airlock_switch.cpp:29]: New file, chance to use references from the start.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	ObjectCollisionBounds LockSwitchBounds =
+	{
+		GameBoundingBox::Zero,
+		std::pair(
+			EulerAngles(ANGLE(-10.0f), ANGLE(-30.0f), ANGLE(-10.0f)),
+			EulerAngles(ANGLE(10.0f), ANGLE(30.0f), ANGLE(10.0f))
+		)
+	};
+
+	auto AirLockSwitchPos = Vector3i::Zero;
```
Comment (Sezzary) [inline TombEngine/Objects/Generic/Switches/airlock_switch.cpp:25]: File elects "airlock" spelling variation, so let's do `Airlock` here as well and for functions.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp (hunk 2 above)
Comment (Adngel) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp:1]: done

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp (hunk 3 above)
Comment (Adngel) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp:32]: done

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp (hunk 4 above)
Comment (Adngel) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp:42]: done

File: TombEngine/Game/control/trigger.cpp (hunk 1 above)
Comment (Adngel) [reply TombEngine/Game/control/trigger.cpp:158]: done

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp (hunk 6 above)
Comment (Adngel) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp:25]: Changed AirLock to Airlock.

I let the capital start because although these, (AirlockSwitchBounds and AirlockSwitchPos) are used only here, they are still global.

File: TombEngine/Objects/Generic/Switches/airlock_switch.cpp (hunk 5 above)
Comment (Adngel) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp:29]: I didn't understood this one.

Are you suggesting to use references instead of pointers?

```
void AirLockSwitchCollision(short itemNumber, ItemInfo& laraItem, CollisionInfo* coll)
{
	auto laraInfo = GetLaraInfo(laraItem);
```

Comment (Sezzary) [reply TombEngine/Objects/Generic/Switches/airlock_switch.cpp]: Yes, I mean that with new stuff we can readily adopt references. So these would be:
`auto& laraInfo = *GetLaraInfo(laraItem);`
`auto& switchItem = g_Level.Items[itemNumber];`

File: TombEngine/Game/control/trigger.cpp [hunk 7]
```diff
@@ -118,48 +118,60 @@ int GetSwitchTrigger(ItemInfo* item, short* itemNos, int attatchedToSwitch)
 	return 0;
 }
 
-int SwitchTrigger(short itemNumber, short timer)
+bool SwitchTrigger(short itemNumber, short timer)
```
Comment (Raildex) [inline TombEngine/Game/control/trigger.cpp:121]: This should not be part of this PR.
Please separate the Airlock Switch from the Trigger Refactor!

**Review by Raildex — CHANGES_REQUESTED** [review]
This PR contains two separate things:
The Air Lock Switch and a refactor for the Triggers.
Please separate the two things for better maintenance later.

File: TombEngine/Game/control/trigger.cpp (hunk 1 above)
Comment (Adngel) [reply TombEngine/Game/control/trigger.cpp:164]: I unmade this change, as Raildex asked, I can remake it on a future branch.

(Once the airlock and switches were merged as both makes some modifications to that function).

File: TombEngine/Game/control/trigger.cpp (hunk 7 above)
Comment (Adngel) [reply TombEngine/Game/control/trigger.cpp]: Yes, Sezz suggested that could be a good chance to do the update, but I neither see any problem postponing that for after the airlock and switches get merged.

I've push another commit that turn the bool back to int as originally was.

File: TombEngine/Game/control/trigger.cpp [hunk 8]
```diff
@@ -118,7 +118,7 @@ int GetSwitchTrigger(ItemInfo* item, short* itemNos, int attatchedToSwitch)
 	return 0;
 }
 
-bool SwitchTrigger(short itemNumber, short timer)
+int SwitchTrigger(short itemNumber, short timer)
```
Comment (Kubsy) [inline TombEngine/Game/control/trigger.cpp:121]: I've seen this in other code files but how about use casting from bool to int? because now if you have "return 1". I mean ik it's obvious but it's for readiness.

**Review by Kubsy — APPROVED** [review]

File: TombEngine/Game/control/trigger.cpp (hunk 8 above)
Comment (Adngel) [reply TombEngine/Game/control/trigger.cpp:121]: Yes, that's one of the things we will do, but Raildex suggest to do that in a different branch, so I unmade the bool update.

**Review by Sezzary — APPROVED** [review]
